### PR TITLE
Forest vs tree

### PR DESCRIPTION
**Pull request by deepakpandita57**
Summary of the changes:

- Support for multiprocessing to run trials in parallel and compression to save storage space
- Support for categorical responses and associated metrics
- Methods used to run experiments reported in the Forest vs Tree paper and the EACL paper
- Some minor changes to replace sklearn mean_squared_error() with root_mean_squared_error() as the former method is deprecated

### CODE REVIEW

File: cat_machine_contest_metrics.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
+
+import numpy as np
+import scipy.stats as st
+from scipy.spatial.distance import jensenshannon
+import sklearn.metrics
+
+def binarize(scores: np.ndarray, threshold: float) -> np.ndarray:
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py:8]: Use docstrings (one line desc is fine) for these funcs like the ones below.

File: cat_machine_contest_metrics.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+  return np.argmax(counts, axis=1)
+
+def cat_accuracy(
+    human: np.ndarray,
+    machine1: np.ndarray,
+    machine2: np.ndarray,
+) -> tuple[float, float]:
+  """Compute accuracy relative to human labels.
+
+  The params ht, mt1 and mt2 can be specified via a config string,
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: Is this comment still relevant? The API only takes 3 separate params.

File: cat_machine_contest_metrics.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    human: A list of human scores.
+    machine1: A list of machine scores.
+    machine2: Another list of machine scores.
+
+  Returns:
+    A pair of accuracy scores, for machines 1 and 2, relative to
+    human scores.
+  """
+
+  num_categories = np.max(np.concatenate((human.flatten(), machine1.flatten(), machine2.flatten()))) + 1
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: Nit: split into multiple lines to avoid exceeding 80 chars.

File: cat_machine_contest_metrics.py (hunk 1 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py:8]: Sure

File: cat_machine_contest_metrics.py (hunk 2 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: You are right. I'll remove this.

File: cat_machine_contest_metrics.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
+
+import numpy as np
+import scipy.stats as st
+from scipy.spatial.distance import jensenshannon
+import sklearn.metrics
+
+def binarize(scores: np.ndarray, threshold: float) -> np.ndarray:
+  """Convert an array of scores into a binary array based on a threshold.
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: float scores in [0, 1]

File: cat_machine_contest_metrics.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    threshold (float): The threshold value for binarization.
+      Scores below this value will be set to 0,
+      and scores equal to or above this value will be set to 1.
+
+  Returns:
+    np.ndarray: The binarized array.
+  """
+  return np.where(scores < threshold, 0, 1)
+
+def freq_agg(arr: np.ndarray, num_categories: int = 3) -> np.ndarray:
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: np.ndarray[Any, np.dtype[np.int_]]

File: cat_machine_contest_metrics.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+  """Convert an array of category labels into an array of category frequencies.
+
+  Args:
+    arr (np.ndarray): An array of integer category labels.
+    num_categories (int, optional): The number of categories. Defaults to 3.
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: nit: why 3 for default? I would have chosen 0 (meaning max+1).

File: cat_machine_contest_metrics.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    lambda x: np.bincount(x, minlength=num_categories),
+    axis=1,
+    arr=arr)
+
+def majority_vote(arr: np.ndarray, num_categories: int = 3) -> np.ndarray:
+  """Convert an array of category labels into an array of
+    majority category labels.
+
+  Args:
+    arr (np.ndarray): An array of integer category labels.
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: Isn't this a matrix?

File: cat_machine_contest_metrics.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+  Returns:
+    np.ndarray: An array of category frequencies for each row in `arr`.
+  """
+  return np.apply_along_axis(
+    lambda x: np.bincount(x, minlength=num_categories),
+    axis=1,
+    arr=arr)
+
+def majority_vote(arr: np.ndarray, num_categories: int = 3) -> np.ndarray:
+  """Convert an array of category labels into an array of
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: Convert matrix of integer categories into array of row-wise pluralities.

File: cat_machine_contest_metrics.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    human: A list of human scores.
+    machine1: A list of machine scores.
+    machine2: Another list of machine scores.
+
+  Returns:
+    A pair of accuracy scores, for machines 1 and 2, relative to
+    human scores.
+  """
+
+  num_categories = np.max(np.concatenate((
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: Why do you need to flatten and concatenate? You should be able to use np.max to compute the global maximum over all scalar values here. Maybe I'm missing something.

File: cat_machine_contest_metrics.py [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+    human.flatten(), machine1.flatten(), machine2.flatten()))) + 1
+  human = majority_vote(human, num_categories)
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: Rename to different var names (say, human_majorities, etc).

File: cat_machine_contest_metrics.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+  machine1 = majority_vote(machine1, num_categories)
+  machine2 = majority_vote(machine2, num_categories)
+
+  return (
+      sklearn.metrics.accuracy_score(human, machine1),
+      sklearn.metrics.accuracy_score(human, machine2),
+  )
+
+def cat_actual_accuracy(
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py:98]: How's this different from cat_accuracy? Docstrings should make clear.

File: cat_machine_contest_metrics.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+
+def cat_auc(
+    human: np.ndarray,
+    machine1: np.ndarray,
+    machine2: np.ndarray,
+) -> tuple[float, float]:
+  """Compute ROC AUC relative to human labels.
+
+  Args:
+    human: A list of human scores.
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: array

File: cat_machine_contest_metrics.py [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+    machine1: A list of machine scores.
+    machine2: Another list of machine scores.
+
+  Returns:
+    A pair of precision scores, for machines 1 and 2, relative to
+    human scores.
+  """
+
+  num_categories = np.max(np.concatenate((
```
Comment (pk-at-g) [inline cat_machine_contest_metrics.py]: It seems like there's much shared intermediate processing for these funcs. Maybe use a shared function?

File: cat_machine_contest_metrics_test.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
+import numpy as np
+from absl.testing import absltest
+import cat_machine_contest_metrics as cm
```
Comment (pk-at-g) [inline cat_machine_contest_metrics_test.py]: Nit: only use library aliases for well known library aliases.

Comment (deepakpandita57) [reply cat_machine_contest_metrics_test.py]: Got it

File: cat_machine_contest_metrics.py (hunk 6 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: Thanks for pointing it out. It was there just because the first dataset I used had 3 categories.

File: cat_machine_contest_metrics.py (hunk 7 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: Yes, it is.

File: cat_machine_contest_metrics.py (hunk 11 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py:98]: It's intended to be calculated based on the categorical parameters directly.

File: cat_machine_contest_metrics.py (hunk 9 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: Yes, I can use stack and max here.

**Comment by tripperroc** [comment]
Hi Deepak,

I’m going to take Samay and Mamadou out for dinner tomorrow, to celebrate submission of our paper. It’s also Mamadou’s birthday and Samay is leaving at the end of the week. Would you like to join us (on me)? We haven’t celebrated our conference attendances (or even acceptances, have we?)

C

From: Deepak Pandita ***@***.***>
Date: Wednesday, February 25, 2026 at 4:46 PM
To: google-research/vet ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [google-research/vet] Forest vs tree (PR #2)

@deepakpandita57 commented on this pull request.

________________________________

In cat_machine_contest_metrics.py<https://github.com/google-research/vet/pull/2#discussion_r2855673024>:

> +    machine1: np.ndarray,

+    machine2: np.ndarray,

+) -> tuple[float, float]:

+  """Compute accuracy relative to human labels.

+

+  Args:

+    human: A list of human scores.

+    machine1: A list of machine scores.

+    machine2: Another list of machine scores.

+

+  Returns:

+    A pair of accuracy scores, for machines 1 and 2, relative to

+    human scores.

+  """

+

+  num_categories = np.max(np.concatenate((

Yes, I can use stack and max here.

File: cat_machine_contest_metrics.py (hunk 5 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: sorry, I don't understand what you mean here

Comment (pk-at-g) [reply cat_machine_contest_metrics.py]: I was suggesting to use a more restrictive type hint, since we know the form of the array here.

File: cat_machine_contest_metrics.py (hunk 6 above)
Comment (pk-at-g) [reply cat_machine_contest_metrics.py]: You should update the docstring accordingly. BTW do you want to require a minlength arg (as it is now) or provide a default value (if so, 0 might make sense).

File: cat_machine_contest_metrics.py (hunk 5 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: Got it!

File: cat_machine_contest_metrics.py (hunk 8 above)
Comment (pk-at-g) [reply cat_machine_contest_metrics.py]: Maybe mention that the majorities are being taken along the cols?

File: categorical_sample.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+)
+_DIR_PRIOR_PROB_DIST_PARAMS = flags.DEFINE_list(
+    "dir_prior_prob_dist_params",
+    None,
+    "Parameters for the probability distribution",
+)
+_COMPUTE_ACTUAL_P_VALUES = flags.DEFINE_boolean(
+    "compute_actual_p_values",
+    False,
+    "If true use categorical params directly to compute metrics and p-values.",
```
Comment (pk-at-g) [inline categorical_sample.py:72]: Give a pointer to the formula that would be used to compute an "actual" (analytic?) p-value.

File: categorical_sample.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+)
+_COMPRESSION = flags.DEFINE_string(
+    "compression",
+    None,
+    "Specifies the compression to use (gz, bz2, xz, or lz4).",
```
Comment (pk-at-g) [inline categorical_sample.py]: Where? For output files?

File: categorical_sample.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+  )
+  
+  if _COMPUTE_ACTUAL_P_VALUES.value:
+    output_filename = os.path.join(
+        _EXP_DIR.value,
+        f"cat_actual_responses_simulated_distr_dist={_DISTORTION.value}_gen_N="
+        f"{_N_ITEMS.value}_K={_K_RESPONSES.value}_M={_M_CATEGORIES.value}"
+        f"_num_samples={_NUM_SAMPLES.value}.{file_extension}",
+    )
+    write_samples_to_file(
```
Comment (pk-at-g) [inline categorical_sample.py]: use full library path here.

File: categorical_sample_lib.py [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+from typing import Any, List, Optional, Tuple
+
+from absl import logging
+import numpy as np
+import datatypes
+import cat_machine_contest_metrics as cmcm
+
+def gen_dirichlet_samples(
+    alpha: List[float],
+    n: int,
```
Comment (pk-at-g) [inline categorical_sample_lib.py]: nit: num_samples

File: categorical_sample_lib.py [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+
+  Args:
+      categorical_params (np.ndarray): Categorical parameters
+      noise_params (np.ndarray): Noise parameters
+      distortion (float): distortion
+
+  Returns:
+      np.ndarray: Distorted parameters
+  """
+  distorted_params = (1 - distortion) * \
```
Comment (pk-at-g) [inline categorical_sample_lib.py:42]: nit: wrap the RHS in parens to enable proper linebreak rather than using backsplash.

File: categorical_sample_lib.py [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+    [rng.multinomial(1, categorical_params).argmax(axis=-1)
+     for _ in range(k_responses)]
+  )
+  responses_z = np.column_stack(
+    [rng.multinomial(1, distorted_params).argmax(axis=-1)
+     for _ in range(k_responses)]
+  )
+  return responses_gold, responses_y, responses_z
+
+def mix_arrays(
```
Comment (pk-at-g) [inline categorical_sample_lib.py:77]: nit: use np.where(random_mask, arr1, arr2) for this. Clarify in docstring what "mix" means and that it's random.

File: categorical_sample_lib.py [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+    [rng.multinomial(1, mix_arrays(distorted_params_null, categorical_params_null)).argmax(
+      axis=-1) for _ in range(k_responses)]
+  )
+  return responses_gold, responses_y, responses_z
+
+def simulate_response_tables_cat(
+    n_items: int = 1000,
+    k_responses: int = 5,
+    m_categories: int = 3,
+    alpha: List[float] = [0.6, 0.1, 0.3],
```
Comment (pk-at-g) [inline categorical_sample_lib.py:133]: Why these default values? Note: they're different from the _ALPHA flag default values.

File: parameterized_sample_lib.py [hunk 22]
```diff
@@ -36,14 +36,58 @@
...
+
+def bi_norm_bounded_sample(
+    mean1: float,
+    std1: float,
+    mean2: float,
+    std2: float,
+    mixture: float,
+    bounds: Tuple[float, float],
+):
+  """Sample from [0,1] via a binormal distribution.
```
Comment (pk-at-g) [inline parameterized_sample_lib.py]: Do you mean sample from [[0, 1], [0, 1]]? It looks like you're sampling from one of two possible univariate normal distributions. Binormal implies bivariate.

File: parameterized_sample_lib.py [hunk 23]
```diff
@@ -259,6 +300,24 @@ def likert_norm_dist(mean: float, std: float, rate: int = 5) -> float:
   x = int(x * rate) / rate
   return x if x < 1 else (rate - 1) / rate
 
+def binary_norm_dist(mean: float, std: float) -> float:
```
Comment (pk-at-g) [inline parameterized_sample_lib.py:307]: Fix the docstring. It seems you're sampling from a normal distribution and then rounding to {0, 1}.

File: categorical_sample.py (hunk 15 above)
Comment (deepakpandita57) [reply categorical_sample.py:72]: This does not refer to the analytic p-values. Here it means that the categorical parameters sampled are directly used for p-value calculation.

File: categorical_sample.py (hunk 16 above)
Comment (deepakpandita57) [reply categorical_sample.py]: Yes, I'll specify in the docstring

File: categorical_sample_lib.py (hunk 21 above)
Comment (deepakpandita57) [reply categorical_sample_lib.py:133]: No particular reason, except this is the fraction of categories in the DICES dataset. I have changed the flag values to match these, but I'm open to your suggestions.

File: parameterized_sample_lib.py (hunk 22 above)
Comment (deepakpandita57) [reply parameterized_sample_lib.py]: You're right. The docstring is not correct. I have fixed it now. I think these are inherited artifacts from an older branch.

File: parameterized_sample_lib.py (hunk 23 above)
Comment (deepakpandita57) [reply parameterized_sample_lib.py:307]: Fixed

File: cat_machine_contest_metrics.py (hunk 6 above)
Comment (deepakpandita57) [reply cat_machine_contest_metrics.py]: I think how it is right now is good since `np.bincount` also has an optional minlength arg
